scripts/calculateRevenue/protocols/aave/blockchainData/oracle.ts

```diff
@@ -0,0 +1,64 @@
+import memoize from '@github/memoize'
+import { Address } from 'viem'
+import BigNumber from 'bignumber.js'
+import { NetworkId } from '../../../../types'
+import { getViemPublicClient } from '../../../../utils'
+import { oracleAbi } from '../../../../abis/aave/oracle'
+
+// Fetches USD prices for a list of tokens using an Aave price oracle
+export const getUSDPrices = memoize(_getUSDPrices, {
+  hash: (...params: Parameters<typeof _getUSDPrices>) => JSON.stringify(params),
+})
+
+export async function _getUSDPrices({
+  networkId,
+  oracleAddress,
+  tokenAddresses,
+  blockNumber,
+}: {
+  networkId: NetworkId
+  oracleAddress: Address
+  tokenAddresses: Address[]
+  blockNumber: number
+}): Promise<Map<Address, BigNumber>> {
+  const publicClient = getViemPublicClient(networkId)
+
+  // Check if the contract exists at the given block number
+  const poolByteCode = await publicClient.getCode({
+    address: oracleAddress,
+    blockNumber: BigInt(blockNumber),
+  })
+
+  const baseCurrencyUnit = poolByteCode
+    ? await publicClient.readContract({
+        address: oracleAddress,
+        abi: oracleAbi,
+        functionName: 'BASE_CURRENCY_UNIT',
+        blockNumber: BigInt(blockNumber),
+      })
+    : 0n
+
+  const prices = poolByteCode
```

Comment:
i wondered if we should parallelise these contract reads, then i remembered we have the batching set on the viem level! cool